### ISSUE TITLE
Semantic Bug in categorical hinge Loss ? 

### ISSUE THREAD

**Issue by unrealwill** [keras-team/keras#8083]
In Wikipedia, the categorical hinge loss is defined as 

![image](https://user-images.githubusercontent.com/11304248/31310265-39bba3f6-ab95-11e7-8c1c-1b06ee23c77c.png)

Please note that the max doesn't include the true label.

In the current code : 
```
def categorical_hinge(y_true, y_pred):
    pos = K.sum(y_true * y_pred, axis=-1)
    neg = K.max((1. - y_true) * y_pred, axis=-1)
    return K.maximum(0., neg - pos + 1.)
```

It seems to work as intended when y_pred are greater than 1 (Please verify it)
But when y_pred are all negative values for example the neg term is "involuntarily "(??) ceiled to 0 instead of being the max of all negative-label.
As an example you can take y_true =[1,0] y_pred = [-3.0,-5.0] the margin is 2 so the loss should be 0 instead the loss is 4.0.

Your formulation has the merit of pushing the true prediction to be greater than 1 (so it drives it towards the domain where it will work OK), but can you confirm that this is the intended behaviour by documenting the source paper for the hinge Loss.

As a correction I would propose to translate the y_pred to the domain where they yield a correct margin calculation.
by adding as a first line 

//Not runned and not tested : 
`y_pred = 1 + y_pred - K.min( y_pred,axis=-1,keepdims=True)`

**Comment by atomextranova** [comment]
I do not this a bug. The implementation is exactly the same as the algorithm described here.[http://www.ttic.edu/sigml/symposium2011/papers/Moore+DeNero_Regularization.pdf](url)

I have not checked your example, but I guess that the algorithm just presume that label is positive.

**Comment by unrealwill** [comment]
Hello

This bug is quite old so I had to refresh myself a bit about it. So please forgive the interrupted train of thought.
I maintain that there is a subtle implementation bug.

I agree with the formula the paper  you quote : 
"Max over r { Mr x  + (1-delta(y,r) }   - My x"
This is the max over a set of element : it means you "pick" the one for which the value over the set is maximum :
Let say the argmax is r1 then Max over r terms is : Mr1 x + (1-delta(y,r1))
Mr1 x  can be negative, delta is 0 or 1 so **this term is unbounded and can go negative.**

and still maintain there is an implementation bug :

`neg = K.max((1. - y_true) * y_pred, axis=-1)`
**This term (which should Mr1 x) can't be negative** because it has excessive "columns" of "0" (because of the 
implementation bug) , so when you take the max it can't go negative because of those .

To be correct The term neg should "pick" the value of row for which the (y_pred + delta) is the max
Because semantically the vectorization of the "pick" is not as easy a simple multiplication, and the neg term seems to do the do the trick we have this bug.

Take my upper example to convince yourself.

Thank you

**Comment by unrealwill** [comment]
//Not runned and not tested :
Correct vectorization should be : 
K.max( y_pred - K.max( y_pred*y_true,axis=-1) )

**Comment by unrealwill** [comment]
@fchollet Adding you in the loop because this 6 months old subtle bug silently messing with all keras SVMs, fixable with a simple (ready but "untested and unrunned") one line fix (see first comment) to check and push, should have already been fixed and not forgotten has it was. Thank you.

**Comment by unrealwill** [comment]
@fchollet ^^Up

**Comment by atomextranova** [comment]
Sorry, I finally see what you really means now. Refer to #2830 for clarification. Here the y_pred and y_true are assumed to be one-hot coded, which is usually recommended and used in standard keras model.

**Comment by unrealwill** [comment]
From memory (untested) the correct code should be something like that

```
def categorical_hinge(y_true, y_pred):
    pos = K.sum(y_true * y_pred, axis=-1)
#Note the inversion of max and maximum when compared to current version
    return K.mean( K.max( K.maximum(0.0, (1.0 - y_true) * y_pred - pos + 1),axis=-1), axis=-1)
```

There is a correctly looking multiclass hinge loss in tensorflow : 
" sparse_multiclass_hinge_loss"
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/kernel_methods/python/losses.py

Naively you can write the correct hinge loss version using some gather ops as : 
```
def hingeLoss(sparseLabel, logit):
    oh = tf.one_hot(sparseLabel, tf.shape(logit)[1])
    pos = tf.gather_nd(logit, tf.where(oh))
    noh = 1 - oh
    neg = tf.reshape(tf.gather_nd(logit, tf.where(noh)), (tf.shape(sparseLabel)[0], -1))
    margin = 1.0
    # Crammer and Singer definition (wikipedia article)
    # loss = tf.reduce_max( tf.nn.relu( margin + neg - tf.expand_dims(pos,axis=1) ) ,axis=1 )
    # Weston and Watkins definition (wikipedia article)
    loss = tf.reduce_sum(tf.nn.relu(margin + neg - tf.expand_dims(pos, axis=1)), axis=1)
    return loss
```

**Comment by atomextranova** [comment]
Thanks for your reply. I do use the current implementation to get similar performance compared to normal cross entropy loss, which means current version should work over one-hot coded samples. I will test your implementation but I do not think it will improve anything at least in my code.

**Comment by Dzugaru** [comment]
Agree with @unrealwill, this non-symmetric loss looks horrible and should be fixed. 

Shouldn't have any visible effect on performance though (negative scores are quickly and uniformly corrected to be positive).

As a side note, if anyone is interested, this is how multiclass hinge loss looks in Caffe (converted to Keras code). First, it uses all false-positive scores, not only the worst at a given moment (which, imo, should help to learn faster and be more stable). Second, it enforces the margin to be centered at 0 (which, imo, should reduce the variance of prediction scores - [1000, 1002] scores are not possible).

```
def categorical_hinge(y_true, y_pred):
    pos = K.sum(y_true * y_pred, axis=-1)
    neg = K.sum((1. - y_true) * y_pred, axis=-1)
    return K.maximum(0., 1. - pos) + K.maximum(0., 1. + neg)
```
